# Rainfall prediction ML model

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv("../data/weather_rainfall_prediction.csv")

In [3]:
df.head()

,Date,Location,MinTemp,MaxTemp,Rainfall,Evaporation,Sunshine,WindGustDir,WindGustSpeed,WindDir9am,...,Humidity9am,Humidity3pm,Pressure9am,Pressure3pm,Cloud9am,Cloud3pm,Temp9am,Temp3pm,RainToday,RainTomorrow
0,2010-04-30,Adelaide,10.8,21.2,0.0,1.8,6.60,ENE,22.0,ENE,...,66.0,48.0,1028.4,1024.7,4.45,4.51,15.7,21.0,No,No
1,2014-07-22,Adelaide,3.7,19.0,0.0,1.4,7.61,N,24.0,N,...,75.0,32.0,1024.2,1021.1,4.45,4.51,9.6,18.5,No,No
2,2014-07-23,Adelaide,9.6,15.8,0.0,2.6,7.61,NE,52.0,NNE,...,35.0,51.0,1017.9,1011.5,4.45,4.51,14.7,15.4,No,Yes
3,2014-07-24,Adelaide,10.1,15.5,16.6,0.8,7.61,NNE,50.0,NW,...,96.0,80.0,1007.4,1008.0,4.45,4.51,11.2,14.0,Yes,Yes
4,2014-07-25,Adelaide,11.2,16.2,1.8,0.6,7.61,NW,30.0,WNW,...,89.0,70.0,1018.4,1019.2,4.45,4.51,13.0,15.3,Yes,Yes


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 23 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Date           145460 non-null  object 
 1   Location       145460 non-null  object 
 2   MinTemp        145460 non-null  float64
 3   MaxTemp        145460 non-null  float64
 4   Rainfall       145460 non-null  float64
 5   Evaporation    145460 non-null  float64
 6   Sunshine       145460 non-null  float64
 7   WindGustDir    145460 non-null  object 
 8   WindGustSpeed  145460 non-null  float64
 9   WindDir9am     145460 non-null  object 
 10  WindDir3pm     145460 non-null  object 
 11  WindSpeed9am   145460 non-null  float64
 12  WindSpeed3pm   145460 non-null  float64
 13  Humidity9am    145460 non-null  float64
 14  Humidity3pm    145460 non-null  float64
 15  Pressure9am    145460 non-null  float64
 16  Pressure3pm    145460 non-null  float64
 17  Cloud9am       145460 non-nul

In [5]:
df.shape

(145460, 23)

In [6]:
df.describe()

,MinTemp,MaxTemp,Rainfall,Evaporation,Sunshine,WindGustSpeed,WindSpeed9am,WindSpeed3pm,Humidity9am,Humidity3pm,Pressure9am,Pressure3pm,Cloud9am,Cloud3pm,Temp9am,Temp3pm
count,145460.000000,145460.000000,145460.000000,145460.000000,145460.000000,145460.000000,145460.000000,145460.000000,145460.000000,145460.000000,145460.000000,145460.000000,145460.000000,145460.000000,145460.000000,145460.000000
mean,12.193993,23.221337,2.360898,5.468995,7.610612,40.035567,14.043384,18.662601,68.880816,51.539143,1017.649946,1015.256314,4.448437,4.509959,16.990624,21.683306
std,6.365750,7.088124,8.382488,3.161541,2.729486,13.118253,8.861059,8.716581,18.854765,20.471189,6.728467,6.663973,2.265605,2.092954,6.449221,6.850057
min,-8.500000,-4.800000,0.000000,0.000000,0.000000,6.000000,0.000000,0.000000,0.000000,0.000000,980.500000,977.100000,0.000000,0.000000,-7.200000,-5.400000
25%,7.700000,18.000000,0.000000,4.000000,7.610000,31.000000,7.000000,13.000000,57.000000,37.000000,1013.500000,1011.100000,3.000000,4.000000,12.300000,16.700000
50%,12.100000,22.700000,0.000000,5.470000,7.610000,39.000000,13.000000,18.660000,69.000000,51.540000,1017.650000,1015.260000,4.450000,4.510000,16.800000,21.400000
75%,16.800000,28.200000,1.000000,5.470000,8.700000,46.000000,19.000000,24.000000,83.000000,65.000000,1021.800000,1019.400000,6.000000,6.000000,21.500000,26.200000
max,33.900000,48.100000,371.000000,145.000000,14.500000,135.000000,130.000000,87.000000,100.000000,100.000000,1041.000000,1039.600000,9.000000,9.000000,40.200000,46.700000


In [7]:
df.columns

Index(['Date', 'Location', 'MinTemp', 'MaxTemp', 'Rainfall', 'Evaporation',
       'Sunshine', 'WindGustDir', 'WindGustSpeed', 'WindDir9am', 'WindDir3pm',
       'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm',
       'Pressure9am', 'Pressure3pm', 'Cloud9am', 'Cloud3pm', 'Temp9am',
       'Temp3pm', 'RainToday', 'RainTomorrow'],
      dtype='object')

## Analysis and next process

Our dataset have `(145460 rows and 23 columns)` and we have zero null values which is a plus point. 
Now our target variable is **`RainTomorrow`** we can also chose `RainToday` but `RainTomorrow` goes more with idea and data.
Now to reduce some complexity we will drop some columns which are not contrubating in our model. And we will combine some columns to reduce it's size, like `Pressure_mean = (Pressure9am + Pressure3pm) / 2` and `Pressure_diff = Pressure3pm − Pressure9am`.

In [8]:
df.columns

Index(['Date', 'Location', 'MinTemp', 'MaxTemp', 'Rainfall', 'Evaporation',
       'Sunshine', 'WindGustDir', 'WindGustSpeed', 'WindDir9am', 'WindDir3pm',
       'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm',
       'Pressure9am', 'Pressure3pm', 'Cloud9am', 'Cloud3pm', 'Temp9am',
       'Temp3pm', 'RainToday', 'RainTomorrow'],
      dtype='object')

In [9]:
# combine preasure meas & prassure difference
df['Pressure_mean'] = (df['Pressure9am'] + df['Pressure3pm'])/2
df['Pressure_diff'] = df['Pressure3pm'] - df['Pressure9am']

In [10]:
# combination of WindDir
df['WindSpeed_mean'] = (df['WindSpeed9am'] + df['WindSpeed3pm'])/2
df['WindSpeed_diff'] = df['WindSpeed3pm'] - df['WindSpeed9am']

In [11]:
# combination of Humidity
df['Humidity_mean'] = (df['Humidity9am'] + df['Humidity3pm'])/2
df['Humidity_diff'] = df['Humidity3pm'] - df['Humidity9am']

In [12]:
# combination of cloud
df['Cloud_mean'] = (df['Cloud9am'] + df['Cloud3pm'])/2
df['Cloud_diff'] = df['Cloud3pm'] - df['Cloud9am']

In [13]:
# combination of Temp
df['Temp_range'] = (df['Temp9am'] + df['Temp3pm'])/2
df['Temp_diff'] = df['Temp3pm'] - df['Temp9am']

In [14]:
df.columns

Index(['Date', 'Location', 'MinTemp', 'MaxTemp', 'Rainfall', 'Evaporation',
       'Sunshine', 'WindGustDir', 'WindGustSpeed', 'WindDir9am', 'WindDir3pm',
       'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm',
       'Pressure9am', 'Pressure3pm', 'Cloud9am', 'Cloud3pm', 'Temp9am',
       'Temp3pm', 'RainToday', 'RainTomorrow', 'Pressure_mean',
       'Pressure_diff', 'WindSpeed_mean', 'WindSpeed_diff', 'Humidity_mean',
       'Humidity_diff', 'Cloud_mean', 'Cloud_diff', 'Temp_range', 'Temp_diff'],
      dtype='object')

Extracting some useful data form `Date` feature

In [15]:
df['Date'] = pd.to_datetime(df['Date'])

In [16]:
# month
df['Month'] = df['Date'].dt.month

In [17]:
df.head()

,Date,Location,MinTemp,MaxTemp,Rainfall,Evaporation,Sunshine,WindGustDir,WindGustSpeed,WindDir9am,...,Pressure_diff,WindSpeed_mean,WindSpeed_diff,Humidity_mean,Humidity_diff,Cloud_mean,Cloud_diff,Temp_range,Temp_diff,Month
0,2010-04-30,Adelaide,10.8,21.2,0.0,1.8,6.60,ENE,22.0,ENE,...,-3.7,8.0,-2.0,57.0,-18.0,4.48,0.06,18.35,5.3,4
1,2014-07-22,Adelaide,3.7,19.0,0.0,1.4,7.61,N,24.0,N,...,-3.1,5.5,11.0,53.5,-43.0,4.48,0.06,14.05,8.9,7
2,2014-07-23,Adelaide,9.6,15.8,0.0,2.6,7.61,NE,52.0,NNE,...,-6.4,14.5,11.0,43.0,16.0,4.48,0.06,15.05,0.7,7
3,2014-07-24,Adelaide,10.1,15.5,16.6,0.8,7.61,NNE,50.0,NW,...,0.6,21.5,9.0,88.0,-16.0,4.48,0.06,12.60,2.8,7
4,2014-07-25,Adelaide,11.2,16.2,1.8,0.6,7.61,NW,30.0,WNW,...,0.8,14.0,2.0,79.5,-19.0,4.48,0.06,14.15,2.3,7


In [18]:
df.drop(columns=['Date', 'WindSpeed9am', 'WindSpeed3pm', 'Humidity9am', 'Humidity3pm',
       'Pressure9am', 'Pressure3pm', 'Cloud9am', 'Cloud3pm', 'Temp9am',
       'Temp3pm', 'RainToday'], inplace=True)

In [19]:
df.shape

(145460, 22)

In [20]:
df.head()

,Location,MinTemp,MaxTemp,Rainfall,Evaporation,Sunshine,WindGustDir,WindGustSpeed,WindDir9am,WindDir3pm,...,Pressure_diff,WindSpeed_mean,WindSpeed_diff,Humidity_mean,Humidity_diff,Cloud_mean,Cloud_diff,Temp_range,Temp_diff,Month
0,Adelaide,10.8,21.2,0.0,1.8,6.60,ENE,22.0,ENE,NW,...,-3.7,8.0,-2.0,57.0,-18.0,4.48,0.06,18.35,5.3,4
1,Adelaide,3.7,19.0,0.0,1.4,7.61,N,24.0,N,NNE,...,-3.1,5.5,11.0,53.5,-43.0,4.48,0.06,14.05,8.9,7
2,Adelaide,9.6,15.8,0.0,2.6,7.61,NE,52.0,NNE,NE,...,-6.4,14.5,11.0,43.0,16.0,4.48,0.06,15.05,0.7,7
3,Adelaide,10.1,15.5,16.6,0.8,7.61,NNE,50.0,NW,NW,...,0.6,21.5,9.0,88.0,-16.0,4.48,0.06,12.60,2.8,7
4,Adelaide,11.2,16.2,1.8,0.6,7.61,NW,30.0,WNW,WNW,...,0.8,14.0,2.0,79.5,-19.0,4.48,0.06,14.15,2.3,7


In [23]:
df['WindDir9am'].value_counts()

WindDir9am
N      22324
SE      9287
E       9176
SSE     9112
NW      8749
S       8659
W       8459
SW      8423
NNE     8129
NNW     7980
ENE     7836
NE      7671
ESE     7630
SSW     7587
WNW     7414
WSW     7024
Name: count, dtype: int64

In [24]:
df['WindDir3pm'].value_counts()

WindDir3pm
SE     15066
W      10110
S       9926
WSW     9518
SSE     9399
SW      9354
N       8890
WNW     8874
NW      8610
ESE     8505
E       8472
NE      8263
SSW     8156
NNW     7870
ENE     7857
NNE     6590
Name: count, dtype: int64

In [25]:
# Converting direction into angle

direction_map = {
    'N': 0, 'NNE': 22.5, 'NE': 45, 'ENE': 67.5,
    'E': 90, 'ESE': 112.5, 'SE': 135, 'SSE': 157.5,
    'S': 180, 'SSW': 202.5, 'SW': 225, 'WSW': 247.5,
    'W': 270, 'WNW': 292.5, 'NW': 315, 'NNW': 337.5
}

df['WindDir9am_angle'] = df['WindDir9am'].map(direction_map)
df['WindDir3pm_angle'] = df['WindDir3pm'].map(direction_map)

In [26]:
df.head()

,Location,MinTemp,MaxTemp,Rainfall,Evaporation,Sunshine,WindGustDir,WindGustSpeed,WindDir9am,WindDir3pm,...,WindSpeed_diff,Humidity_mean,Humidity_diff,Cloud_mean,Cloud_diff,Temp_range,Temp_diff,Month,WindDir9am_angle,WindDir3pm_angle
0,Adelaide,10.8,21.2,0.0,1.8,6.60,ENE,22.0,ENE,NW,...,-2.0,57.0,-18.0,4.48,0.06,18.35,5.3,4,67.5,315.0
1,Adelaide,3.7,19.0,0.0,1.4,7.61,N,24.0,N,NNE,...,11.0,53.5,-43.0,4.48,0.06,14.05,8.9,7,0.0,22.5
2,Adelaide,9.6,15.8,0.0,2.6,7.61,NE,52.0,NNE,NE,...,11.0,43.0,16.0,4.48,0.06,15.05,0.7,7,22.5,45.0
3,Adelaide,10.1,15.5,16.6,0.8,7.61,NNE,50.0,NW,NW,...,9.0,88.0,-16.0,4.48,0.06,12.60,2.8,7,315.0,315.0
4,Adelaide,11.2,16.2,1.8,0.6,7.61,NW,30.0,WNW,WNW,...,2.0,79.5,-19.0,4.48,0.06,14.15,2.3,7,292.5,292.5


In [28]:
df.drop(columns=['WindDir9am', 'WindDir3pm'], inplace = True)

In [29]:
df.head()

,Location,MinTemp,MaxTemp,Rainfall,Evaporation,Sunshine,WindGustDir,WindGustSpeed,RainTomorrow,Pressure_mean,...,WindSpeed_diff,Humidity_mean,Humidity_diff,Cloud_mean,Cloud_diff,Temp_range,Temp_diff,Month,WindDir9am_angle,WindDir3pm_angle
0,Adelaide,10.8,21.2,0.0,1.8,6.60,ENE,22.0,No,1026.55,...,-2.0,57.0,-18.0,4.48,0.06,18.35,5.3,4,67.5,315.0
1,Adelaide,3.7,19.0,0.0,1.4,7.61,N,24.0,No,1022.65,...,11.0,53.5,-43.0,4.48,0.06,14.05,8.9,7,0.0,22.5
2,Adelaide,9.6,15.8,0.0,2.6,7.61,NE,52.0,Yes,1014.70,...,11.0,43.0,16.0,4.48,0.06,15.05,0.7,7,22.5,45.0
3,Adelaide,10.1,15.5,16.6,0.8,7.61,NNE,50.0,Yes,1007.70,...,9.0,88.0,-16.0,4.48,0.06,12.60,2.8,7,315.0,315.0
4,Adelaide,11.2,16.2,1.8,0.6,7.61,NW,30.0,Yes,1018.80,...,2.0,79.5,-19.0,4.48,0.06,14.15,2.3,7,292.5,292.5


In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145460 entries, 0 to 145459
Data columns (total 22 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Location          145460 non-null  object 
 1   MinTemp           145460 non-null  float64
 2   MaxTemp           145460 non-null  float64
 3   Rainfall          145460 non-null  float64
 4   Evaporation       145460 non-null  float64
 5   Sunshine          145460 non-null  float64
 6   WindGustDir       145460 non-null  object 
 7   WindGustSpeed     145460 non-null  float64
 8   RainTomorrow      145460 non-null  object 
 9   Pressure_mean     145460 non-null  float64
 10  Pressure_diff     145460 non-null  float64
 11  WindSpeed_mean    145460 non-null  float64
 12  WindSpeed_diff    145460 non-null  float64
 13  Humidity_mean     145460 non-null  float64
 14  Humidity_diff     145460 non-null  float64
 15  Cloud_mean        145460 non-null  float64
 16  Cloud_diff        14